### ISSUE TITLE
[Feature]: Workspace Key/Value storage  "Extensions"

### ISSUE THREAD

**Issue by magliok-wwt** [appsmithorg/appsmith#37398]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Summary

I would like to have a way to set static variables at the workspace, environment, and app level within AppSmith.
The idea is this would be a way to do key : value storage for static values needed within apps.

The concept is similar to another tool we utilize and these 'extensions' work very well for these use-cases.
- We need to provide a link to another application ( appsmith and non-appsmith ) and the URL is different in dev/test/prod
- We have links to images/assets that need to change per environment
- We need to change wording based on the environment

A future state of this could also function similarly to a 'resources' strings type feature that many apps see as extensibility to support multiple languages. This function functions similarly in that you can generate keys and provide values. The only change to this feature would be to support setting the language/local for the values as you populate them ( and being able to reference what language you are selecting when retrieving said values ) -- localization should be a Phase 2.

We need the ability to have a Key - become an Extension.
These Keys can be setup/configured within an App.

Once the app is published, there will be an area in the AppSmith admin UI where values can be assigned.
These extensions would - show up only after the app is published, therefore you only specify the value.
This reduces having to "coordinate" knowing/typing in the exact 'key' again.

When setting these values, you should be able to set them at either the Environment ( within a Workspace ) level - optionally having a setting for "all" environments ( so effectively you are setting this at a Workspace level ).

App: My App
 - extension 'variable' :  serviceNowURL

Publish "My App" to the Production workspace

Workspace Settings Gear > Extensions 

This would then list extensions (Keys), grouped by App Name
You would hit a '+' button to add an extension
You are asked "One value for all Environments" or "Value per Environment"
From here, you either get to enter the one value if it's for all environments or a way to enter values for each environment configured for the workspace.

Should a new environment be added later, it should be expected that you have to set those values - if you are using environment-specific configurations Extension(s).

### Why should this be worked on?

This is a must-have to be able to correctly specify different assets when you operate AppSmith in a multi-environment enterprise. This should be enhanced to allow multi-lingual support as well... as this could be done utilizing a similar feature.  (Environment within a workspace configurable strings).

We need a solution, without hard-coding these values into applications.
Today, there isn't a great solution to target Workspace / Environment names in JS, this could rely on that or circumvent that need. 

I don't know how a mature shop, interconnecting Apps built in AppSmith, could reference links that change across environments in any other way without this feature.

**Comment by ame-appsmith** [comment]
[EE ticket](https://internal.appsmith.com/app/oracle/ticket-details-662d74ffc3e4d34941e56ede?zid=3311)

**Comment by magliok-wwt** [comment]
@Nikhil-Nandagopal - No they would not

1. Packages have to be deployed on every single page that they are needed, they are not application-wide 
2. Packages still have the problem, meaning we'd have to manage figure out what environment or workspace it's within.
3. We are not going to develop "live" in production - meaning maintain a package "code" only in production, as we full intend to utilize git integration up-to and including CI-CD.

Side note - packages are an okay start; they need some love to be a really good feature, to be honest.

**Comment by magliok-wwt** [comment]
That's the thing @hendrawanhadikusuma  - The only solution they're talking about is hardcoding properties of a JS object in a package.  The problem is having different 'environments' and having different variables for each.  

Today there is no way to target the App/Workspace you're within ( from JS ) and therefore you can't even make a switch/if statements to adjust an 'environment variable' to each environment. 

You'd have to build and manage a package - directly within each environment/workspace -- e.g. live develop in production.

**Comment by hendrawanhadikusuma** [comment]
@magliok-wwt actually something i aware is there have URL params called `environment` but it not always appear

**Comment by Nikhil-Nandagopal** [comment]
We will look into adding the environment variable as a context variable in the editor and I believe that + packages should solve this

**Comment by magliok-wwt** [comment]
hey @Nikhil-Nandagopal  - I believe there is another similar request I think you linked it. 

But if we can get the App, Workspace, and Environment ( at minimum ) added to the JS object' app smith', then yes, with packages, we can make a version of this. 

I need all three ( App, Workspace, Environment ), but I would like to have the option to have one universal package ( for every app ) for this (call it "Settings").

**Comment by Nikhil-Nandagopal** [comment]
Yes we're adding all 3 and you can connect the package to git and import it on every workspace to treat it as a universal settings package

**Comment by magliok-wwt** [comment]
Awesome

That other ticket was
[26610](https://github.com/appsmithorg/appsmith/issues/26610)

You can see my comment - I'd love to get more contextual items - but those three will help for sure.
